File: Python/AP-1/merge_two.py

# start with two arrays of strings, A and B, each with its elements in alphabetical order and without duplicates
# return a new array containing the first N elements from the two arrays
# the result array should be in alphabetical order and without duplicates
# A and B will both have a length which is N or more
def merge_two(a, b, n):
    new_list = []
    one = 0
    two = 0

    for _ in range(n):
        if a[one] < b[two]:
            new_list.append(a[one])
            one += 1
        elif a[one] > b[two]:
            new_list.append(b[two])
            two += 1
        else:
            new_list.append(a[one])
            one += 1
            two += 1

    return new_list

print(merge_two(['a', 'c', 'z'], ['b', 'f', 'z'], 3))
print(merge_two(['a', 'c', 'z'], ['c', 'f', 'z'], 3))
print(merge_two(['f', 'g', 'z'], ['c', 'f', 'g'], 3))

File: JavaScript/AP-1/mergeTwo.js

// start with two arrays of strings, A and B, each with its elements in alphabetical order and without duplicates
// return a new array containing the first N elements from the two arrays
// the result array should be in alphabetical order and without duplicates
// A and B will both have a length which is N or more
const mergeTwo = (a, b, n) => {
    let newArray = [];
    let count = 0;
    let index1 = 0;
    let index2 = 0;

    for (let i = 0; i < a.length + b.length; i++) {
        if (count === n) {
            break;
        } else {
            if (a[index1] < b[index2]) {
                newArray.push(a[index1]);
                count++;
                index1++;
            } else if (a[index1] > b[index2]) {
                newArray.push(b[index2]);
                count++;
                index2++;
            } else {
                newArray.push(a[index1]);
                count++;
                index1++;
                index2++;
            }
        }
    }

    return newArray;
};

console.log(mergeTwo(['a', 'c', 'z'], ['b', 'f', 'z'], 3));
console.log(mergeTwo(['a', 'c', 'z'], ['c', 'f', 'z'], 3));
console.log(mergeTwo(['f', 'g', 'z'], ['c', 'f', 'g'], 3));
